File: JavaScript/find-common-letters.js

// Find common letters of 2 given strings

function commonLetters (s1,s2) {
    const letters1 = {}
    s1.split('').reduce((freq,character)=> {
        freq[character] = -~freq[character];
        return freq;
    },letters1);

    const letters2 = {}
    s2.split('').reduce((freq,character)=> {
        freq[character] = -~freq[character];
        return freq;
    },letters2); 

    let ch;
    for(var i =0; i <=25; i++ ) {
        ch = String.fromCharCode(i + 97);
        if(letters1[ch] && letters2[ch]) {
            console.log(ch) // a, d, m, s
        }
    }

}

console.log(1)
let s1 = "asdajdasm";
let s2 = "dasmkmkqsk";
commonLetters(s1,s2);

File: C++/Diverse/find-common-letters.cpp


// Find common letters of 2 strings

#include <iostream>
#include <cstring>
using namespace std;

void commonLetters(char s1[], char s2[]) {

    // Use 2 vector to count the occurrences
    int *fr1 = new int[26];
    int *fr2 = new int[26];

    // First string
    for(int i  = 0; i < strlen(s1);i++) {
        // ~undefined = -1
        // ~1 = -2; -~1 = 2 and so on..
        fr1[s1[i] - 97] = -~fr1[s1[i] - 97];
    }

    // Second string
    for(int i  = 0; i < strlen(s2);i++) {
        fr2[s2[i] - 97] = -~fr2[s2[i] - 97];
    }

    for(int i =0; i <= 25;i++) {
        // If there is a common letter
        if(fr1[i] && fr2[i]) {
            cout << char(i + 97) <<" ";
        }
    }
}

int main (){
    
    char s1 [] = "asdajdasm" ;
    char s2 [] = "dasmkmkqsk";
    commonLetters(s1,s2); // a d m s
    return 0;
}   